File: 05_22_2024_Minimize_Max_Distance_to_Gas_Station.cpp
class Solution {
  public:
    int ok(double x, vector<int> &v) {
        int ret = 0;
        for (int i = 0; i < v.size() - 1; i++) {
            ret += ceil((v[i + 1] - v[i]) / x) - 1;
        }
        return ret;
    }

    double findSmallestMaxDist(vector<int> &stations, int K) {
        double low = 0;
        int n = stations.size();
        double high = stations[n - 1] - stations[0];
        while (high - low >= 1e-6) {
            double mid = (low + high) / 2.0;
            int x = ok(mid, stations);
            if (x > K) {
                low = mid;
            } else {
                high = mid;
            }
        }
        return high;
    }
};

File: 05_22_2024_Minimize_Max_Distance_to_Gas_Station.py
class Solution:

    # Function to find the smallest maximum distance
    def findSmallestMaxDist(self, stations, K):
        left, right = 1e-6, stations[-1] - stations[0]
        while left + 1e-6 < right:
            mid = (left + right) / 2
            count = 0

            # Counting the number of stations that can be added with the current maximum distance
            for a, b in zip(stations, stations[1:]):
                count += math.ceil((b - a) / mid) - 1

            # Updating the left or right values based on the count
            if count > K:
                left = mid
            else:
                right = mid
        return right
